### ISSUE TITLE
More flexible answer generation

### ISSUE THREAD

**Issue by whitead** [jerryjliu/gpt_index#440]
Really like the interface and very impressed with the quick progress on the package! I've been exploring different approaches to using the raw value returned from documents in question answering. I did some experiments and found that instead of the refining approach, it works better for me to do a "hierarchical" approach and summarize and then combine. [See overview here](https://github.com/whitead/paper-qa).

I did find a way to get the raw sources out via this approach:

```py
query = gpt_index.indices.query.vector_store.simple.GPTSimpleVectorIndexQuery(
                        index.index_struct,
                        similarity_top_k=k,
                        embed_model=index.embed_model,
                        prompt_helper=lambda: None,  # dummy
                    )
queries[index] = query
query = queries[index]
nodes = query.get_nodes_and_similarities_for_response(question)
```

This was fine, but I couldn't figure out how to store the metadata I wanted in the nodes too. Maybe this is me missing something in the docs, but I wanted to store page numbers in the metadata to use later on. I couldn't figure out how to do that though.

## TL;DR

So here's what I would propose:

- [ ] Have two paths for answering: one based on refinement, one based on summarizing from multiple sources [see prompts](https://github.com/whitead/paper-qa/blob/main/paperqa/qaprompts.py)
- [ ] Enable metadata to be stored with documents that can be access in prompt templates
- [ ] Make it possible to easily get nodes - without instantiating query obj
- [ ] [minor] Expose max_marginal for FAISS (maybe already done?) - which I found improves diversity of responses

These are just my observations in my use-case, it seems that gpt-index has been very successful for others. 
I'm happy to help write code on this - in fact I tried to - but this seems to involve some deep rethinking of library. So - happy to discuss and contribute where I can - but will need help from gpt-index team.

**Comment by emptycrown** [comment]
I think this is a great idea! Will let @jerryjliu address the best way to PR, but specifically regarding the metadata, you can include it using `extra_info` when creating a `Document`. So like:
```python
doc = Document(text, extra_info={"page number": 1})
```

For example, [SimpleDirectoryReader](https://github.com/emptycrown/llama-hub/blob/main/loader_hub/file/base.py) uses this to enable you to pass in arbitrary metadata for files. This metadata is then prepended to each node if the document is split up. The [ArxivReader](https://github.com/emptycrown/llama-hub/blob/main/loader_hub/papers/arxiv/base.py) uses it to pass in things like the title of the paper, the author, date published, etc.

**Comment by whitead** [comment]
Great thanks! I actually want to format the metadata myself in the prompt - maybe this is part of the larger changes; but I couldn't figure out how to format the metadata as I wanted. It just seems to appear in the prompt.

**Comment by emptycrown** [comment]
Ah yeah more fine-grained customization there would be really useful for sure. Looking forward to your PR :)

**Comment by jerryjliu** [comment]
@whitead not sure if this helps, but following up on jesse's point, you can set `include_extra_info=False` when initializing the index. This way, you can still propagate metadata from documents to nodes to source nodes in the response, but it won't be injected into each node. Then you could still format the metadata yourself in the prompt (i think).

**Comment by whitead** [comment]
That solves part of my problem @jerryjliu! Do you think it's possible to have the node text and metadata be returned for a query, without all the boilerplate code above? Would that be something I can PR if not?

**Comment by jerryjliu** [comment]
@whitead if i'm understanding correctly you could just fetch source_nodes from the response

```
response = index.query(...)
source_nodes = response.source_nodes
for source_node in source_nodes:
   # print node text and document metadata
   print((source_node.source_text, source_node.extra_info))
```

**Comment by whitead** [comment]
Got it - thanks @jerryjliu for helping me understand this. Does that still generate an LLM call from the query, aside from the query embedding?

**Comment by jerryjliu** [comment]
> Got it - thanks @jerryjliu for helping me understand this. Does that still generate an LLM call from the query, aside from the query embedding?

oops oh you can also disable the LLM call by setting response_mode="no_text" during the `index.query` call - that way we don't generate an answer at all, we only return the source nodes

**Comment by whitead** [comment]
Perfect - this addresses my use case. I can implement the other points in my own code.
